lib/Modules/ComputeHooks/ComputeHooks.psm1

```diff
@@ -108,972 +96,437 @@ function Install-Prerequisites {
-                },
-                @{
-                    "generator"="Get-S2DContext";
-                    "relation"="s2d";
-                }
-                );
-        };
-        "neutron-ovs"=@{
-            "myname"="neutron-ovs";
-            "template"="$template_dir\$distro\ml2_conf.ini"
-            "service"="neutron-openvswitch-agent";
-            "binpath"="$neutronOpenvswitchExe";
-            "serviceBinPath"="`"$serviceWrapperNeutron`" neutron-openvswitch-agent `"$neutronOpenvswitchExe`" --config-file `"$neutron_ml2`"";
-            "config"="$neutron_ml2";
-            "context_generators"=@(
-                @{
-                    "generator"="Get-RabbitMQContext";
-                    "relation"="amqp";
-                },
-                @{
-                    "generator"="Get-NeutronContext";
-                    "relation"="cloud-compute";
-                },
-                @{
-                    "generator"="Get-CharmConfigContext";
-                    "relation"="config";
-                },
-                @{
-                    "generator"="Get-SystemContext";
-                    "relation"="system";
-                },
-                @{
-                    "generator"="Get-S2DContext";
-                    "relation"="s2d";
-                }
-                );
-        };
+    foreach ($context in $service['context_generators']) {
+        Write-JujuInfo ("Getting context for {0}" -f $context["relation"])
+        $allContexts.Add($context["relation"])
+        $ctx = & $context["generator"]
+        Write-JujuInfo ("Got {0} context: {1}" -f @($context["relation"], $ctx.Keys))
+        if (!$ctx.Count -and $context["mandatory"]) {
+            # Context is empty. Probably peer not ready
+            Write-JujuWarning ("Context for {0} is EMPTY" -f $context["relation"])
+            $incompleteContexts.Add($context["relation"])
+            $shouldRestart = $false
+            continue
+        }
+        foreach ($val in $ctx.Keys) {
+            $mergedContext[$val] = $ctx[$val]
+        }
+    }
+    Set-IncompleteStatusContext -ContextSet $allContexts -Incomplete $incompleteContexts
+    if (!$mergedContext.Count) {
+        return $false
     }
-    return $JujuCharmServices
+    Start-RenderTemplate -Context $mergedContext -TemplateName $service["template"] -OutFile $service["config"]
+    return $shouldRestart
 }
 
-function Get-RabbitMQContext {
-    Write-JujuLog "Generating context for RabbitMQ"
-    $username = Get-JujuCharmConfig -Scope 'rabbit-user'
-    $vhost = Get-JujuCharmConfig -Scope 'rabbit-vhost'
-    if (!$username -or !$vhost){
-        Write-JujuError "Missing required charm config options: rabbit-user or rabbit-vhost"
+function Get-DataPort {
+    # try and set up bonding early. This will create
+    # a new Net-LbfoTeam and try to acquire an IP address
+    # via DHCP. This interface may receive os-data-network IP.
+    $bondName = New-BondInterface
```

Comment:
Which other providers? Juju only supports MAAS as bare metal deployment system (where this code makes sense). Ironic is not yet supported. Do you see this code being used on virtual systems? If so, skip the TODO :).